### ISSUE TITLE
bookdown support

### ISSUE THREAD

**Issue by theGreatWhiteShark** [polymode/polymode#141]
Hey,

I was just playing around with Yihui Xie's [bookdown](https://bookdown.org/yihui/bookdown/) and I love it. It would be really awesome to use it with polymode too!

So, changing/adding some lines to use the export function of the bookdown package is one thing. But it provides a lot more features than markdown itself, like latex-style equations, cross references etc.  https://bookdown.org/yihui/bookdown/markdown-extensions-by-bookdown.html 

In other words: the syntax highlighting of the markdown-mode is just not powerful enough and it would call for a **bookdown-mode**. 

Unfortunately I never wrote a package for emacs so it would most probably take me quite some time to implement it all by myself.

Is someone interested in implementing this mode? Or in helping me to do so?

**Comment by brendan-r** [comment]
I'd find it useful too. However it's worth being specific about what bookdown support would actually mean. 

**Number and reference equations**
It seems that good syntax highlighting for the code blocks would be covered by an existing LaTeX mode. There's already font-locking for text between `$` marks as part of `markdown-mode`. I don't write too much math so I've never configured polymode to trigger LaTeX syntax highlighting, but this *seems* possible.

**Theorems and proofs**
Here it seems all we'd have to do is make 
```
```{theorem}`
```
... and the other keywords in table 2.1 trigger a LaTeX mode block. This sounds achievable via polymode configuration.

**Special headers**
This is a potentially useful feature, but I don't think I'd need `markdown-mode` to do anything differently here?

**Text references**
This is the best part of bookdown for me. It seems like support here might mean:
- Font locking (colorized text) for strings that look like `\(ref\:*\)`
- *Possibly* autocompletion of them at `(ref:|)`, but this isn't a big deal


Do you think that would cover it?

**Comment by vspinu** [comment]
What export command would be needed? Isn't it automatically picked by Rmarkdown from the yaml header?

**Comment by theGreatWhiteShark** [comment]
@brendan-r Yes, these are the main ones.

And it's true, `markdown-mode` itself provides highlighting of LaTeX-style equations via the `markdown-enable-math` variable. But this is just pleasant for inline equations.
When writing **block equations** using 
``` math
$$\begin{equation}
  L(\theta) = f(x_1,x_2, \ldots, x_n | \theta )
  (\#eq:reference)
\end{equation}$$
```
the double dollar sign `$$` does highlight the code and prevents backslashes and underscores from doing silly things, but the highlighting is in single, plain red. Not even symbols highlighted differently in LaTeX, like `//` or `&`, do have another face. 

This and the missing highlighting of the text references are my main concerns. 

In addition I also plan to disable auto-fill-mode for all lines containing a heading. The reference label `{...}` is getting pushed down and misinterpreted way to often.

For the **theorem** environment I'm not really sure if I'm getting the point. For me it's basically just a bunch of text and equations. If both of them are lighted properly, it would be fine by me.

The status of the mode right now: I started digging into the Emacs reference and into the code of `markdown-mode`. But there is still loads of things I have to understand and implement. I guess in two weeks I might have the first version done, which tries to take care of the highlighting.

**Comment by theGreatWhiteShark** [comment]
@vspinu The big problem I see with `bookdown` is that its content is very likely to be scattered across multiple documents. In the root folder you have a file *_bookdown.yml* specifying a shared header including a list of all documents belonging to the book.

> rmd_files: ["01-introduction.Rmd", "02-chapter.Rmd", "03-chapter.Rmd", "06-summary.Rmd"]

When triggering the `polymode` export in any of these files only the file itself is going to be exported. 

Setting 

> output: bookdown::git_document

in the YAML header will trigger the export to the desired format but, again, just for the file visited.

**Comment by vspinu** [comment]
I see. But then the book could be huge, you don't want to re-compile the whole book, right? I guess bookdown should be doing some smart caching behind and re-compile only changed files. In that case, adding a dedicated exporter would be useful indeed.

**Comment by theGreatWhiteShark** [comment]
Yes, `bookdown` caches all code chunks of the book separately and seems to check for all included files whether their chunks have been touched.

**Comment by brendan-r** [comment]
@vspinu Any tips on getting polymode to trigger a latex chunk/block/buffer, within markdown, mode via `$$` delimted lines?

I've tried reading your docs and messing around with things like the following, but unfortunately I didn't get very far.

```elisp
(defcustom  pm-inner/markdown+latex
  (pm-hbtchunkmode "markdown+latex"
                   :mode 'latex-mode
                   :head-reg  "^[ \t]*\\$\\$[ \t]*$"
                   :tail-reg  "^[ \t]*\\$\\$[ \t]*$")
  "This should trigger latex mode, maybe?"
  :group 'innermodes
  :type 'object)


(defcustom pm-poly/markdown+latex
  (pm-polymode-one "markdown+latex"
                   :hostmode 'pm-host/markdown
                   :innermode 'pm-inner/markdown+latex)
  "A LaTeX block inside a markdown document, maybe"
  :group 'polymodes
  :type 'object)

(define-polymode poly-markdown+latex-mode pm-poly/markdown+latex)


;; A made-up file-extension for testing
(add-to-list 'auto-mode-alist '("\\.mdm" . poly-markdown+latex-mode))

```

**Comment by vspinu** [comment]
@brendan-r I have just added support for latex blocks to markdown [here](https://github.com/vspinu/polymode/blob/next/modes/poly-markdown.el#L63-L71). It's in the `next` branch and it is essentially adding one more inner block to the blocks that markdown understands.

The inner mode currently starts the block if $$ is on a new line (white spaces allowed) and ends the block with $$ before the new line (possibly preceded by white spaces). 

Unfortunately developing (and even maintaning) polymode at this moment is out of my reach. Nor I have time to merge `next` branch at the moment. Sorry about that.  You will have to pull `next` branch yourself for now.

**Comment by student-sx** [comment]
I just checked out the `next` branch and tried it. It works for blocks of the form:

```
$$
\int_a^b f(x) dx $$
```

but not for 

```
Some text

$$
\int_a^b f(x) dx 
$$

Some text
```

Since `%` is seen as comment by latex, you could also write something like:

````
$$%
\int_a^b f(x) dx 
%$$

````

One seems to need define a coding style which breaks the symmetry. In my second example this would be a blank line followed by `$$` followed by a non blank line for the starting of the environment and a non blank line followed by `$$` followed by a blank line for the end of the environment. Could make this style also work? I don't see the regular expression to use.  Maybe it would be nice to add some code block examples (as above) to the documentation for other users.

**Comment by student-sx** [comment]
By the way: Why not also parse inline math: $\int_a^b f(x) dx$: Pattern: Whitespace then $ then non-whitespace (but not $) for start and non-whitespace (but not $), $ and whitespace for end?

**Comment by vspinu** [comment]
> One seems to need define a coding style which breaks the symmetry.

Indeed. One  needs to differentiate between head and tail; otherwise one would need to parse the text from the beginning of the buffer each time. I like the comment idea. it's enough to have it at the end. What is the bookdown specification for  the head latex chunk?

How about imposing that head $$ should always be followed by something (can be space or comment) and tail should be followed by a newline? I have just pushed a change which does exactly that.

> Why not also parse inline math.

Can be, but isn't this specific to bookdown only? Do any other markdown parsers understand latex at all? 

 I would favor creating a dedicated mode derived from markdown for that. Then one could add all those other goodies mentioned by @brendan-r without affecting other users of markdown.

**Comment by theGreatWhiteShark** [comment]
> but isn't this specific to bookdown only? Do any other markdown parsers understand latex at all?

As far as I do understand the *bookdown* package it is just a wrapper around *rmarkdown*. It takes care of the concatenation of all files, the caching, and a number of additional parsing of the final document to add some new flavors. The actual markdown parser is **pandoc** and is able to handle LaTeX type inline math.

In classical .Rmd documents rendered using *rmarkdown* you can include inline math using a slightly different, MathJax-type, syntax `\(\int_a^b f(x) dx\)`. For some reason the *rmarkdown* package can not handle the `$` delimiter but pandoc itself does https://pandoc.org/MANUAL.html#math

**Comment by vspinu** [comment]
Ok. I think it makes sense to support pandoc specs. I have just added inline-math innermode to markdown.

**Comment by vspinu** [comment]
I just realized that the inline math is not that simple. In some language leading dollars mean something and dollars could occur freely within strings. A good solution in this case seems to be restrict inline span to top levels only.

**Comment by theGreatWhiteShark** [comment]
Branch `next`: commit  0deaca0

Polymode does not properly recognizes ESS-mode chunks whenever it is preceded by one of the custom chunks of the bookdown package.

````

```{theorem 'theorem-1'}

A theorem	
		
```

```{r}
## Polymode enables FallBack and not ESS mode in here
x <- 1

```
````

Instead of ESS-mode, **FallBack** will be detected for both chunks.

**Comment by vspinu** [comment]
Works just fine for me on your example. I don't see why `thorem` should be any more special than any other name. If you do `el`, `r` or `blabla` instead of `theorem` does it work?

**Comment by theGreatWhiteShark** [comment]
The bug is persistent under changing the chunk name from *theorem* to *blabla*. Changing it to *r* or *el* fixes the issue. 

When executing `kill-buffer` with point in either the r or theorem chunk in the above example I get the following error message.

> (poly-lock-fontify-now 1 120 [span 53 116 minimal.Rmd[r]]) -> (font-lock-default-fontify-region 53 116): Selecting deleted buffer

Afterwards the second chunk is correctly identified as ESS.

But the error message above occurs over and over again and I am only able to kill the buffer when point is outside of either the theorem or r chunk.

(BTW disabling `yasnippet` won't fix the issue this time)

**Comment by vspinu** [comment]
I actually see the highlighting issue, but not the error which you describe. You are on Emacs 27 right?

> When executing kill-buffer 

Interesting. Never happened to me. I guess font lock is in progress when you delete the buffer. I will make a guard for this. 


Please open new issues you notice something astray. It's easier to track things that way.

**Comment by vspinu** [comment]
Both issues should be fixed now, but I am not entirely sure about the buffer-kill one. I just placed a few safeguards here and there.

 Thanks!

**Comment by theGreatWhiteShark** [comment]
Yes, the recognition of ESS-mode in the chunk works quite fine again. The issue with `kill-buffer` nevertheless persists. I'll open a new issue.

**Comment by vspinu** [comment]
The core of this issue is addressed, for follow ups please open new issues on poly-R or poly-markdown trackers.
